tools/pkg/engine_build_configs/lib/src/build_config_runner.dart

```diff
@@ -0,0 +1,553 @@
+  final bool runGenerators;
+
+  /// Whether to run the test step. Defaults to true.
+  final bool runTests;
+
+  @override
+  Future<bool> run(RunnerEventHandler eventHandler) async {
+    if (!build.canRunOn(platform)) {
+      eventHandler(RunnerError(
+        build.name, <String>[], DateTime.now(),
+        'Build with drone_dimensions "{${build.droneDimensions.join(',')}}" '
+        'cannot run on platform ${platform.operatingSystem}',
+      ));
+      return false;
+    }
+
+    if (runGn) {
+      if (!await _runGN(eventHandler)) {
+        return false;
+      }
+    }
+
+    if (runNinja) {
+      if (!await _runNinja(eventHandler)) {
+        return false;
+      }
+    }
+
+    if (runGenerators) {
+      if (!await _runGenerators(eventHandler)) {
+        return false;
+      }
+    }
+
+    if (runTests) {
+      if (!await _runTests(eventHandler)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  // GN arguments from the build config that can be overridden by extraGnArgs.
+  static const List<(String, String)> _overridableArgs = <(String, String)>[
+    ('--lto', '--no-lto'),
+    ('--rbe', '--no-rbe'),
+    ('--goma', '--no-goma'),
+  ];
+
+  // extraGnArgs overrides the build config args.
+  late final Set<String> _mergedGnArgs = () {
+    // Put the union of the build config args and extraGnArgs in gnArgs.
+    final Set<String> gnArgs = Set<String>.of(build.gn);
+    gnArgs.addAll(extraGnArgs);
+
+    // If extraGnArgs contains an arg, remove its opposite from gnArgs.
+    for (final (String, String) arg in _overridableArgs) {
+      if (extraGnArgs.contains(arg.$1)) {
+        gnArgs.remove(arg.$2);
+      }
+      if (extraGnArgs.contains(arg.$2)) {
+        gnArgs.remove(arg.$1);
+      }
+    }
+
+    return gnArgs;
+  }();
+
+  Future<bool> _runGN(RunnerEventHandler eventHandler) async {
+    final String gnPath = p.join(engineSrcDir.path, 'flutter', 'tools', 'gn');
+    final Set<String> gnArgs = _mergedGnArgs;
+    final List<String> command = <String>[gnPath, ...gnArgs];
+    eventHandler(RunnerStart('${build.name}: GN', command, DateTime.now()));
+    final ProcessRunnerResult processResult;
+    if (dryRun) {
+      processResult = _dryRunResult;
+    } else {
+      processResult = await processRunner.runProcess(
+        command,
+        workingDirectory: engineSrcDir,
+        failOk: true,
+      );
+    }
+    final RunnerResult result = RunnerResult(
+      '${build.name}: GN', command, DateTime.now(), processResult,
+    );
+    eventHandler(result);
+    return result.ok;
+  }
+
+  // TODO(zanderso): This should start and stop RBE when it is an --rbe build.
+  Future<bool> _runNinja(RunnerEventHandler eventHandler) async {
+    final String ninjaPath = p.join(
+      engineSrcDir.path, 'flutter', 'third_party', 'ninja', 'ninja',
+    );
+    final String outDir = p.join(engineSrcDir.path, 'out', build.ninja.config);
+    final List<String> command = <String>[
+      ninjaPath,
```

Comment:
Should this use autoninja if GOMA and RBE are not enabled? From my understanding ninja uses a single job by default.

Reply:
No. `autoninja` includes logic that is specific to the Chrome tree, and which doesn't do the right thing in the flutter/engine tree. `ninja` by itself already chooses a good `-j` value. See `ninja -h` to see the default it chooses.

Reply:
Oh wow. I've been using `autoninja` for 2 years. Good to know 😅

Thanks for the explanation, marking as resolved!